File: task_team_manager_backend/app_mobile/models/notification.model.js
const mongoose = require("mongoose");

const notificationSchema = new mongoose.Schema({
  employee_id: {
    type: String,
    ref: "Employee",
    required: true,
  },

  actor_id: {
    type: String,
    ref: "Employee",
    required: true,
  },

  task_id: {
    type: String,
    ref: "Task",
    required: true,
  },

  type: {
    type: String,
    required: true,
    enum: [
      // Task notifications
      "task_assigned",
      "task_updated",
      "task_completed",
      "task_confirmed",
      "task_comment",
      "task_deadline_near",
      "task_overdue",
      // Role notifications
      "role_update",
      "role_assigned",
      // Project notifications
      "project_assigned",
      "project_updated",
      "project_completed",
      "project_deadline_near",
      // Employee notifications
      "employee_created",
      "employee_updated",
      "employee_role_changed",
      "employee_deleted",
    ],
  },

  message: {
    type: String,
    required: true,
  },

  isRead: {
    type: Boolean,
    default: false,
  },

  read_at: {
    type: Date,
    default: null,
  },

  create_at: {
    type: Date,
    default: Date.now,
    required: true,
  },

  //Thêm thông tin bổ sung
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {},
  },
});
const Notification = mongoose.model("Notification", notificationSchema);
module.exports = Notification;

File: task_team_frontend_mobile/lib/models/notification_model.dart
class NotificationModel {
  final String? id;
  final String message;
  final String type;
  final bool isRead;
  final DateTime createAt;
  final DateTime? readAt;
  final Map<String, dynamic>? metadata;

  NotificationModel({
    this.id,
    required this.message,
    required this.type,
    required this.isRead,
    required this.createAt,
    this.readAt,
    this.metadata,
  });

  factory NotificationModel.fromJson(Map<String, dynamic> json) {
    return NotificationModel(
      id: json['_id'] ?? json['id'],
      message: json['message'] ?? '',
      type: json['type'] ?? 'default',
      isRead: json['isRead'] ?? false,
      createAt: DateTime.parse(json['create_at'] ??
          json['createdAt'] ??
          DateTime.now().toIso8601String()),
      readAt: json['read_at'] != null || json['readAt'] != null
          ? DateTime.parse(json['read_at'] ?? json['readAt'])
          : null,
      metadata: json['metadata'] as Map<String, dynamic>?,
    );
  }

  NotificationModel copyWith({
    String? id,
    String? message,
    String? type,
    bool? isRead,
    DateTime? createAt,
    DateTime? readAt,
    Map<String, dynamic>? metadata,
  }) {
    return NotificationModel(
      id: id ?? this.id,
      message: message ?? this.message,
      type: type ?? this.type,
      isRead: isRead ?? this.isRead,
      createAt: createAt ?? this.createAt,
      readAt: readAt ?? this.readAt,
      metadata: metadata ?? this.metadata,
    );
  }
}
